python/paddle/distributed/passes/auto_parallel_recompute.py

```diff
@@ -0,0 +1,388 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import logging
+
+from .pass_base import PassBase, register_pass
+from paddle.fluid import core, unique_name
+from paddle.fluid import framework as framework
+from paddle.fluid.framework import Variable, Operator
+from paddle.fluid.backward import _append_grad_suffix_, _get_no_grad_set_name
+from paddle.fluid.backward import ProgramStats, _rename_arg_, _find_op_path_
+from paddle.distributed.auto_parallel.process_mesh import ProcessMesh
+from paddle.distributed.auto_parallel.dist_attribute import OperatorDistributedAttribute
+from paddle.distributed.auto_parallel.utils import get_loss_op, set_var_dist_attr, set_dist_op_desc_original_id
+from paddle.distributed.auto_parallel.utils import naive_set_dist_op_attr_for_program_by_mesh_and_mapping
+
+
+class RecomputeState(ProgramStats):
+    def __init__(self, block, ops):
+        super(RecomputeState, self).__init__(block=block, ops=ops)
+        self._block = block
+        self._ops = ops
+        self.var_op_deps = {}
+
+    def build_stats(self):
```

Comment:
`build_stats` is `build_state`?

Reply:
This function inherits from `ProgramStats` of `backward.py`.